File: LeetCode/21_MergeTwoSortedLists.cc
// https://leetcode.com/problems/merge-two-sorted-lists/
//  2021 11.23 2022 3.10 3.27


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if(not list1) return list2;
        if(not list2) return list1;
        
        ListNode* a, *b, *ret, *tmp;
        if (list1->val<=list2->val){
            a=list1;
            b=list2;
        }else{
            a=list2;
            b=list1;
        }
        ret=a;
        while (a->next){
            if(a->next->val <= b->val){
                a=a->next;
            }else{
                tmp=a->next;
                a->next=b;
                b=tmp;
            }
        }
        a->next=b;
        return ret;

        // // v2 recursive
        // if(not list1)return list2;
        // if(not list2)return list1;
        // ListNode *small=list1,*big=list2;
        // if(list1->val>list2->val){
        //     small=list2;
        //     big=list1;
        // }
        // small->next=mergeTwoLists(small->next,big);
        
        // return small;
        




    }
};

File: LeetCode/21_MergeTwoSortedLists.go
// https://leetcode.com/problems/merge-two-sorted-lists/
// 2022年07月24日 21:48:47

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 import "fmt"
 func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
     if list1==nil{
         return list2
     }else if list2==nil {
         return list1
     } 
     if list1.Val<list2.Val {
         ret:=list1
         ret.Next=mergeTwoLists(list1.Next, list2)
         return ret
     }else if list1.Val>=list2.Val {
         ret:=list2
         ret.Next=mergeTwoLists(list1, list2.Next)
         return ret
     }
     return nil
     
     
     // // v1 recursive failed to handle void value
     // var ret *ListNode
     // if list1==nil && list2==nil{
     //     return ret
     // }else if list2==nil || list1.Val<list2.Val {
     //     ret:=list1
     //     fmt.Println("list1",ret)
     //     ret.Next=mergeTwoLists(list1.Next, list2)
     //     return ret
     // }else if list1==nil || list1.Val>=list2.Val {
     //     ret:=list2
     //     fmt.Println("list2",ret)
     //     ret.Next=mergeTwoLists(list1, list2.Next)
     //     return ret
     // }
     // fmt.Println("ret",ret)
     // return ret
     
     
     
     
 }